src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs

```diff
@@ -943,62 +987,17 @@ internal virtual void PrepareSession()
                 WebSession.RetryIntervalInSeconds = RetryIntervalSec;
             }
         }
-        
+
         internal virtual HttpClient GetHttpClient(bool handleRedirect)
         {
-            HttpClientHandler handler = new();
-            handler.CookieContainer = WebSession.Cookies;
-            handler.AutomaticDecompression = DecompressionMethods.All;
-
-            // Set the credentials used by this request
-            if (WebSession.UseDefaultCredentials)
-            {
-                // The UseDefaultCredentials flag overrides other supplied credentials
-                handler.UseDefaultCredentials = true;
-            }
-            else if (WebSession.Credentials is not null)
-            {
-                handler.Credentials = WebSession.Credentials;
-            }
-
-            if (NoProxy)
-            {
-                handler.UseProxy = false;
-            }
-            else if (WebSession.Proxy is not null)
-            {
-                handler.Proxy = WebSession.Proxy;
-            }
-
-            if (WebSession.Certificates is not null)
-            {
-                handler.ClientCertificates.AddRange(WebSession.Certificates);
-            }
-
-            if (SkipCertificateCheck)
-            {
-                handler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
-                handler.ClientCertificateOptions = ClientCertificateOption.Manual;
-            }
-
-            // This indicates GetResponse will handle redirects.
-            if (handleRedirect || WebSession.MaximumRedirection == 0)
-            {
-                handler.AllowAutoRedirect = false;
-            }
-            else if (WebSession.MaximumRedirection > 0)
+            HttpClient client = WebSession.GetHttpClient(handleRedirect, TimeoutSec, out bool clientWasReset);
```

Comment:
Can we evaluate `handleRedirect`internally in WebSession?
Can we save TimeoutSec in WebSession too?

Reply:
We can evaluate the TimeoutSec in WebSession. I have made this change.

The handleRedirect flag is not so obvious. It is controlled by a lot of different factors, including a Http Header being present in the current request that we cannot handle in WebSession without adding a lot of duplicated properties too. I think this would likely be more complicated and confusing than the current approach.

Unfortunately, turning on or off automatic (.NET) redirection support means we need to create a new HttpClientHandler and we lose the persistent connections. A possible future PR would be to have PowerShell always do the redirect handling itself - in that way it wouldn't matter what redirect settings were tweaked between invocations on a session. It could use the same handler for requests (so long as other properties are not changed).

@CarloToso & @iSazonov  - what do you think of this approach?

Reply:
I think it should work but it might be slower than the current implementation (currently PowerShell [with handleRedirect==true] is slower than .NET when handling redirects). I don't think we should do it, at least not at the moment. Instead if you want you could try to work on solving this 3 issues #12249, #16122, #16145 (they are closely related to eachother).

Reply:
I think it's okay to break the persistence occasionally for unusual scenarios where redirect handling is differing between requests. It will still be considerably better than it was before with no persistence for any connections for most users.

I may have a look at those issues, but not while this one is still in progress.

Reply:
Yes, my suggestion was for a future PR.